libcaf_io/src/instance.cpp

```diff
@@ -253,230 +231,278 @@ void instance::write_server_handshake(execution_unit* ctx, buffer_type& out_buf,
-                                    const node_id& dest_node, actor_id aid) {
+void instance::write_monitor_message(execution_unit* ctx, buffer_type& buf,
+                                     const node_id& dest_node, actor_id aid) {
   CAF_LOG_TRACE(CAF_ARG(dest_node) << CAF_ARG(aid));
-  header hdr{message_type::announce_proxy, 0, 0, 0,
-             this_node_, dest_node, invalid_actor_id, aid};
-  write(ctx, buf, hdr);
+  auto writer = make_callback([&](serializer& sink) -> error {
+    return sink(this_node_, dest_node);
+  });
+  header hdr{message_type::monitor_message, 0, 0, 0, invalid_actor_id, aid};
+  write(ctx, buf, hdr, &writer);
 }
 
-void instance::write_kill_proxy(execution_unit* ctx, buffer_type& buf,
-                                const node_id& dest_node, actor_id aid,
-                                const error& rsn) {
+void instance::write_down_message(execution_unit* ctx, buffer_type& buf,
+                                  const node_id& dest_node, actor_id aid,
+                                  const error& rsn) {
   CAF_LOG_TRACE(CAF_ARG(dest_node) << CAF_ARG(aid) << CAF_ARG(rsn));
   auto writer = make_callback([&](serializer& sink) -> error {
-    return sink(const_cast<error&>(rsn));
+    return sink(this_node_, dest_node, rsn);
   });
-  header hdr{message_type::kill_proxy, 0, 0, 0,
-             this_node_, dest_node, aid, invalid_actor_id};
+  header hdr{message_type::down_message, 0, 0, 0, aid, invalid_actor_id};
   write(ctx, buf, hdr, &writer);
 }
 
-void instance::write_heartbeat(execution_unit* ctx,
-                               buffer_type& buf,
-                               const node_id& remote_side) {
-  CAF_LOG_TRACE(CAF_ARG(remote_side));
-  header hdr{message_type::heartbeat, 0, 0, 0,
-             this_node_, remote_side, invalid_actor_id, invalid_actor_id};
+void instance::write_heartbeat(execution_unit* ctx, buffer_type& buf) {
+  CAF_LOG_TRACE("");
+  header hdr{message_type::heartbeat, 0, 0, 0, invalid_actor_id,
+             invalid_actor_id};
   write(ctx, buf, hdr);
 }
 
 bool instance::handle(execution_unit* ctx, connection_handle hdl, header& hdr,
                       std::vector<char>* payload) {
-  // function object for checking payload validity
-  auto payload_valid = [&]() -> bool {
-    return payload != nullptr && payload->size() == hdr.payload_len;
-  };
-  // handle message to ourselves
+  CAF_LOG_TRACE(CAF_ARG(hdl) << CAF_ARG(hdr));
+  // Check payload validity.
+  if (payload == nullptr) {
+    if (hdr.payload_len != 0) {
+      CAF_LOG_WARNING("invalid payload");
+      return false;
+    }
+  } else if (hdr.payload_len != payload->size()) {
+    CAF_LOG_WARNING("invalid payload");
+    return false;
+  }
+  // Dispatch by message type.
   switch (hdr.operation) {
     case message_type::server_handshake: {
+      // Deserialize payload.
+      binary_deserializer bd{ctx, *payload};
+      node_id source_node;
+      std::vector<std::string> app_ids;
```

Comment:
Simply because `config_value` does not supports sets at the moment. However, even if we wouldn't have this limitation I would probably have used a vector anyway. No one in his right mind is going to configure thousands of applications IDs, so a vector is more efficient regardless of O(n) lookup due to a compact memory layout and fewer heap allocations.